File: frontend/src/stores/tasks.js
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { apiService } from '@/services/api'

export const useTasksStore = defineStore('tasks', () => {
  // Estado
  const tasks = ref([])
  const currentTask = ref(null)
  const loading = ref(false)
  const error = ref(null)
  const dashboardStats = ref(null)
  const priorities = ref([])

  // Getters
  const getTaskById = computed(() => {
    return (id) => tasks.value.find(task => task.id === id)
  })

  const completedTasks = computed(() => {
    return tasks.value.filter(task => task.completed)
  })

  const pendingTasks = computed(() => {
    return tasks.value.filter(task => !task.completed)
  })

  const overdueTasks = computed(() => {
    return tasks.value.filter(task => task.is_overdue)
  })

  const tasksByPriority = computed(() => {
    const grouped = {}
    priorities.value.forEach(priority => {
      grouped[priority.value] = tasks.value.filter(task => task.priority === priority.value)
    })
    return grouped
  })

  const totalTasks = computed(() => tasks.value.length)

  // Actions
  async function fetchTasks(filters = {}) {
    loading.value = true
    error.value = null
    
    try {
      const response = await apiService.get('/api/tasks', { params: filters })
      
      tasks.value = response.data.tasks || []
      return response.data
    } catch (err) {
      error.value = err.response?.data?.error || 'Erro ao carregar tarefas'
      console.error('Erro ao buscar tarefas:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  async function fetchTaskById(id) {
    loading.value = true
    error.value = null
    
    try {
      const response = await apiService.get(`/api/tasks/${id}`)
      
      currentTask.value = response.data.task
      
      // Atualizar na lista também
      const index = tasks.value.findIndex(task => task.id === id)
      if (index !== -1) {
        tasks.value[index] = response.data.task
      }
      
      return response.data.task
    } catch (err) {
      error.value = err.response?.data?.error || 'Erro ao carregar tarefa'
      console.error('Erro ao buscar tarefa:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  async function createTask(taskData) {
    loading.value = true
    error.value = null
    
    try {
      const response = await apiService.post('/api/tasks', taskData)
      
      const newTask = response.data.task
      tasks.value.unshift(newTask)
      
      return newTask
    } catch (err) {
      error.value = err.response?.data?.error || 'Erro ao criar tarefa'
      console.error('Erro ao criar tarefa:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  async function updateTask(id, updates) {
    loading.value = true
    error.value = null
    
    try {
      const response = await apiService.put(`/api/tasks/${id}`, updates)
      
      const updatedTask = response.data.task
      
      // Atualizar na lista
      const index = tasks.value.findIndex(task => task.id === id)
      if (index !== -1) {
        tasks.value[index] = updatedTask
      }
      
      // Atualizar currentTask se for a mesma
      if (currentTask.value?.id === id) {
        currentTask.value = updatedTask
      }
      
      return updatedTask
    } catch (err) {
      error.value = err.response?.data?.error || 'Erro ao atualizar tarefa'
      console.error('Erro ao atualizar tarefa:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  async function deleteTask(id) {
    loading.value = true
    error.value = null
    
    try {
      await apiService.delete(`/api/tasks/${id}`)
      
      // Remover da lista
      tasks.value = tasks.value.filter(task => task.id !== id)
      
      // Limpar currentTask se for a mesma
      if (currentTask.value?.id === id) {
        currentTask.value = null
      }
      
      return true
    } catch (err) {
      error.value = err.response?.data?.error || 'Erro ao deletar tarefa'
      console.error('Erro ao deletar tarefa:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  async function toggleTaskCompletion(id) {
    loading.value = true
    error.value = null
    
    try {
      const response = await apiService.patch(`/api/tasks/${id}/toggle`)
      
      const updatedTask = response.data.task
      
      // Atualizar na lista
      const index = tasks.value.findIndex(task => task.id === id)
      if (index !== -1) {
        tasks.value[index] = updatedTask
      }
      
      // Atualizar currentTask se for a mesma
      if (currentTask.value?.id === id) {
        currentTask.value = updatedTask
      }
      
      return updatedTask
    } catch (err) {
      error.value = err.response?.data?.error || 'Erro ao alterar status da tarefa'
      console.error('Erro ao toggle tarefa:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  async function fetchPriorities() {
    try {
      const response = await apiService.get('/api/tasks/priorities')
      priorities.value = response.data.priorities || []
      return response.data.priorities
    } catch (err) {
      console.error('Erro ao buscar prioridades:', err)
      // Prioridades padrão como fallback
      priorities.value = [
        { value: 'low', label: 'Baixa', color: '#4caf50' },
        { value: 'medium', label: 'Média', color: '#ff9800' },
        { value: 'high', label: 'Alta', color: '#f44336' },
        { value: 'urgent', label: 'Urgente', color: '#9c27b0' }
      ]
      return priorities.value
    }
  }

  async function fetchDashboardStats() {
    loading.value = true
    error.value = null
    
    try {
      const response = await apiService.get('/api/tasks/dashboard')
      
      dashboardStats.value = response.data.stats
      return response.data
    } catch (err) {
      error.value = err.response?.data?.error || 'Erro ao carregar estatísticas'
      console.error('Erro ao buscar dashboard:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  async function bulkOperation(taskIds, operation, options = {}) {
    loading.value = true
    error.value = null
    
    try {
      const payload = {
        task_ids: taskIds,
        operation,
        ...options
      }
      
      const response = await apiService.post('/api/tasks/bulk', payload)
      
      // Recarregar tarefas após operação em lote
      await fetchTasks()
      
      return response.data
    } catch (err) {
      error.value = err.response?.data?.error || 'Erro na operação em lote'
      console.error('Erro em operação bulk:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  function clearCurrentTask() {
    currentTask.value = null
  }

  function clearError() {
    error.value = null
  }

  function reset() {
    tasks.value = []
    currentTask.value = null
    loading.value = false
    error.value = null
    dashboardStats.value = null
  }

  // Filtros úteis
  function filterTasks(filters) {
    let filtered = [...tasks.value]
    
    if (filters.completed !== undefined) {
      filtered = filtered.filter(task => task.completed === filters.completed)
    }
    
    if (filters.priority) {
      filtered = filtered.filter(task => task.priority === filters.priority)
    }
    
    if (filters.task_list_id) {
      filtered = filtered.filter(task => task.task_list_id === filters.task_list_id)
    }
    
    if (filters.search) {
      const search = filters.search.toLowerCase()
      filtered = filtered.filter(task => 
        task.title.toLowerCase().includes(search) ||
        task.description?.toLowerCase().includes(search)
      )
    }
    
    if (filters.overdue_only) {
      filtered = filtered.filter(task => task.is_overdue)
    }
    
    return filtered
  }

  return {
    // Estado
    tasks,
    currentTask,
    loading,
    error,
    dashboardStats,
    priorities,
    
    // Getters
    getTaskById,
    completedTasks,
    pendingTasks,
    overdueTasks,
    tasksByPriority,
    totalTasks,
    
    // Actions
    fetchTasks,
    fetchTaskById,
    createTask,
    updateTask,
    deleteTask,
    toggleTaskCompletion,
    fetchPriorities,
    fetchDashboardStats,
    bulkOperation,
    clearCurrentTask,
    clearError,
    reset,
    filterTasks
  }
}) 

File: backend/app/routes/tasks.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from datetime import datetime
from app import db
from app.models.task import Task
from app.models.task_list import TaskList

tasks_bp = Blueprint('tasks', __name__)


@tasks_bp.route('/', methods=['GET'])
@jwt_required()
def get_tasks():
    """Listar todas as tarefas do usuário"""
    try:
        current_user_id = get_jwt_identity()
        
        # Parâmetros de query
        task_list_id = request.args.get('task_list_id', type=int)
        completed = request.args.get('completed')
        priority = request.args.get('priority')
        search = request.args.get('search', '').strip()
        overdue_only = request.args.get('overdue_only', 'false').lower() == 'true'
        
        # Query base
        query = Task.query.filter_by(user_id=current_user_id)
        
        # Filtros
        if task_list_id:
            query = query.filter_by(task_list_id=task_list_id)
        
        if completed is not None:
            is_completed = completed.lower() == 'true'
            query = query.filter_by(completed=is_completed)
        
        if priority:
            query = query.filter_by(priority=priority)
        
        if search:
            query = query.filter(
                Task.title.ilike(f'%{search}%') |
                Task.description.ilike(f'%{search}%')
            )
        
        # Executar query
        tasks = query.all()
        
        # Filtrar vencidas (não pode ser feito direto no SQL facilmente)
        if overdue_only:
            tasks = [task for task in tasks if task.is_overdue()]
        
        # Ordenar
        tasks.sort(key=lambda t: (
            t.completed,  # Não concluídas primeiro
            -{'urgent': 4, 'high': 3, 'medium': 2, 'low': 1}.get(t.priority, 0),  # Por prioridade
            t.due_date or datetime.max  # Por data de vencimento
        ))
        
        return jsonify({
            'tasks': [task.to_dict() for task in tasks],
            'total': len(tasks)
        })
        
    except Exception as e:
        return jsonify({'error': 'Erro interno do servidor'}), 500


@tasks_bp.route('/', methods=['POST'])
@jwt_required()
def create_task():
    """Criar nova tarefa"""
    try:
        current_user_id = get_jwt_identity()
        data = request.get_json()
        
        # Validações
        if not data.get('title'):
            return jsonify({'error': 'Título é obrigatório'}), 400
        
        if len(data['title']) > 200:
            return jsonify({'error': 'Título deve ter no máximo 200 caracteres'}), 400
        
        task_list_id = data.get('task_list_id')
        if not task_list_id:
            return jsonify({'error': 'Lista de tarefas é obrigatória'}), 400
        
        # Verificar se a lista pertence ao usuário
        task_list = TaskList.query.filter_by(
            id=task_list_id,
            user_id=current_user_id
        ).first()
        
        if not task_list:
            return jsonify({'error': 'Lista de tarefas não encontrada'}), 404
        
        # Validar prioridade
        valid_priorities = [p['value'] for p in Task.get_priorities()]
        priority = data.get('priority', 'medium')
        if priority not in valid_priorities:
            return jsonify({'error': 'Prioridade inválida'}), 400
        
        # Validar data de vencimento
        due_date = None
        if data.get('due_date'):
            try:
                due_date = datetime.fromisoformat(data['due_date'].replace('Z', '+00:00'))
            except ValueError:
                return jsonify({'error': 'Formato de data inválido'}), 400
        
        # Criar nova tarefa
        task = Task(
            title=data['title'],
            description=data.get('description', ''),
            priority=priority,
            due_date=due_date,
            task_list_id=task_list_id,
            user_id=current_user_id
        )
        
        db.session.add(task)
        db.session.commit()
        
        return jsonify({
            'message': 'Tarefa criada com sucesso',
            'task': task.to_dict()
        }), 201
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Erro interno do servidor'}), 500


@tasks_bp.route('/<int:task_id>', methods=['GET'])
@jwt_required()
def get_task(task_id):
    """Obter tarefa específica"""
    try:
        current_user_id = get_jwt_identity()
        
        task = Task.query.filter_by(
            id=task_id,
            user_id=current_user_id
        ).first()
        
        if not task:
            return jsonify({'error': 'Tarefa não encontrada'}), 404
        
        return jsonify({'task': task.to_dict()})
        
    except Exception as e:
        return jsonify({'error': 'Erro interno do servidor'}), 500


@tasks_bp.route('/<int:task_id>', methods=['PUT'])
@jwt_required()
def update_task(task_id):
    """Atualizar tarefa"""
    try:
        current_user_id = get_jwt_identity()
        data = request.get_json()
        
        task = Task.query.filter_by(
            id=task_id,
            user_id=current_user_id
        ).first()
        
        if not task:
            return jsonify({'error': 'Tarefa não encontrada'}), 404
        
        # Validações
        if 'title' in data:
            if not data['title']:
                return jsonify({'error': 'Título é obrigatório'}), 400
            
            if len(data['title']) > 200:
                return jsonify({'error': 'Título deve ter no máximo 200 caracteres'}), 400
            
            task.title = data['title']
        
        if 'description' in data:
            task.description = data['description']
        
        if 'priority' in data:
            valid_priorities = [p['value'] for p in Task.get_priorities()]
            if data['priority'] not in valid_priorities:
                return jsonify({'error': 'Prioridade inválida'}), 400
            task.priority = data['priority']
        
        if 'due_date' in data:
            if data['due_date']:
                try:
                    task.due_date = datetime.fromisoformat(data['due_date'].replace('Z', '+00:00'))
                except ValueError:
                    return jsonify({'error': 'Formato de data inválido'}), 400
            else:
                task.due_date = None
        
        if 'task_list_id' in data:
            # Verificar se a nova lista pertence ao usuário
            new_list = TaskList.query.filter_by(
                id=data['task_list_id'],
                user_id=current_user_id
            ).first()
            
            if not new_list:
                return jsonify({'error': 'Lista de tarefas não encontrada'}), 404
            
            task.task_list_id = data['task_list_id']
        
        if 'completed' in data:
            if data['completed'] != task.completed:
                task.toggle_completion()
        
        db.session.commit()
        
        return jsonify({
            'message': 'Tarefa atualizada com sucesso',
            'task': task.to_dict()
        })
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Erro interno do servidor'}), 500


@tasks_bp.route('/<int:task_id>', methods=['DELETE'])
@jwt_required()
def delete_task(task_id):
    """Deletar tarefa"""
    try:
        current_user_id = get_jwt_identity()
        
        task = Task.query.filter_by(
            id=task_id,
            user_id=current_user_id
        ).first()
        
        if not task:
            return jsonify({'error': 'Tarefa não encontrada'}), 404
        
        db.session.delete(task)
        db.session.commit()
        
        return jsonify({'message': 'Tarefa deletada com sucesso'})
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Erro interno do servidor'}), 500


@tasks_bp.route('/<int:task_id>/toggle', methods=['PATCH'])
@jwt_required()
def toggle_task_completion(task_id):
    """Marcar/desmarcar tarefa como concluída"""
    try:
        current_user_id = get_jwt_identity()
        
        task = Task.query.filter_by(
            id=task_id,
            user_id=current_user_id
        ).first()
        
        if not task:
            return jsonify({'error': 'Tarefa não encontrada'}), 404
        
        task.toggle_completion()
        db.session.commit()
        
        status = 'concluída' if task.completed else 'reaberta'
        
        return jsonify({
            'message': f'Tarefa {status} com sucesso',
            'task': task.to_dict()
        })
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Erro interno do servidor'}), 500


@tasks_bp.route('/priorities', methods=['GET'])
def get_priorities():
    """Obter lista de prioridades disponíveis"""
    return jsonify({'priorities': Task.get_priorities()})


@tasks_bp.route('/dashboard', methods=['GET'])
@jwt_required()
def get_dashboard_stats():
    """Obter estatísticas para dashboard"""
    try:
        current_user_id = get_jwt_identity()
        
        # Tarefas do usuário
        all_tasks = Task.query.filter_by(user_id=current_user_id).all()
        
        # Estatísticas gerais
        total_tasks = len(all_tasks)
        completed_tasks = len([t for t in all_tasks if t.completed])
        pending_tasks = total_tasks - completed_tasks
        overdue_tasks = len([t for t in all_tasks if t.is_overdue()])
        
        # Tarefas por prioridade
        priority_stats = {}
        for priority_info in Task.get_priorities():
            priority = priority_info['value']
            priority_tasks = [t for t in all_tasks if t.priority == priority and not t.completed]
            priority_stats[priority] = len(priority_tasks)
        
        # Tarefas recentes (últimos 7 dias)
        from datetime import timedelta
        week_ago = datetime.utcnow() - timedelta(days=7)
        recent_tasks = [t for t in all_tasks if t.created_at >= week_ago]
        
        # Próximas tarefas (próximos 7 dias)
        next_week = datetime.utcnow() + timedelta(days=7)
        upcoming_tasks = [
            t for t in all_tasks 
            if t.due_date and not t.completed and t.due_date <= next_week
        ]
        upcoming_tasks.sort(key=lambda t: t.due_date)
        
        return jsonify({
            'stats': {
                'total_tasks': total_tasks,
                'completed_tasks': completed_tasks,
                'pending_tasks': pending_tasks,
                'overdue_tasks': overdue_tasks,
                'completion_rate': round((completed_tasks / total_tasks * 100) if total_tasks > 0 else 0, 2),
                'priority_breakdown': priority_stats,
                'recent_tasks_count': len(recent_tasks),
                'upcoming_tasks_count': len(upcoming_tasks)
            },
            'upcoming_tasks': [task.to_dict() for task in upcoming_tasks[:5]]  # Próximas 5
        })
        
    except Exception as e:
        return jsonify({'error': 'Erro interno do servidor'}), 500


@tasks_bp.route('/bulk', methods=['POST'])
@jwt_required()
def bulk_operations():
    """Operações em lote nas tarefas"""
    try:
        current_user_id = get_jwt_identity()
        data = request.get_json()
        
        task_ids = data.get('task_ids', [])
        operation = data.get('operation')
        
        if not task_ids or not operation:
            return jsonify({'error': 'IDs das tarefas e operação são obrigatórios'}), 400
        
        # Buscar tarefas do usuário
        tasks = Task.query.filter(
            Task.id.in_(task_ids),
            Task.user_id == current_user_id
        ).all()
        
        if len(tasks) != len(task_ids):
            return jsonify({'error': 'Algumas tarefas não foram encontradas'}), 404
        
        # Executar operação
        if operation == 'complete':
            for task in tasks:
                if not task.completed:
                    task.toggle_completion()
        
        elif operation == 'incomplete':
            for task in tasks:
                if task.completed:
                    task.toggle_completion()
        
        elif operation == 'delete':
            for task in tasks:
                db.session.delete(task)
        
        elif operation == 'move':
            new_list_id = data.get('target_list_id')
            if not new_list_id:
                return jsonify({'error': 'Lista de destino é obrigatória'}), 400
            
            # Verificar se a lista pertence ao usuário
            target_list = TaskList.query.filter_by(
                id=new_list_id,
                user_id=current_user_id
            ).first()
            
            if not target_list:
                return jsonify({'error': 'Lista de destino não encontrada'}), 404
            
            for task in tasks:
                task.task_list_id = new_list_id
        
        else:
            return jsonify({'error': 'Operação inválida'}), 400
        
        db.session.commit()
        
        return jsonify({
            'message': f'Operação {operation} executada em {len(tasks)} tarefa(s)',
            'affected_count': len(tasks)
        })
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Erro interno do servidor'}), 500 